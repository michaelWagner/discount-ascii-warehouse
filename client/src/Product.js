import React from 'react';
import ProductGrid from './ProductGrid';
import ProductAttribute from './ProductAttribute';

class Product extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (<tr>
      {this.props.product.map((productAttribute, idx) => (
        <ProductAttribute productAttribute={productAttribute} />
      ))}
    </tr>);
  }
}
export default Product;
