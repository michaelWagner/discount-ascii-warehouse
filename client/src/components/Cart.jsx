import React from 'react';
import Product from './Product';

class Cart extends React.Component {

  render() {
    const cartProducts = this.props.products.map((product, idx) => {
      return (
        <tbody key={idx}>
          <Product key={product.id} toggleProductInCart={this.props.toggleProductInCart}
                   product={product} formatPriceFromCentsToDollars={this.props.formatPriceFromCentsToDollars}
                   parentComponent={"Cart"}/>
        </tbody>
      );
    });

    return (
      <table className={ (this.props.cartVisible ? "cart-table" : "cart-hidden") + " product-grid-table"}>
        <tfoot className="cart-table-footer">
          <tr className="table-footer">
            <td colSpan="2">Total price: </td>
            <td colSpan="4">{this.props.formatPriceFromCentsToDollars(this.props.totalPrice)}</td>
          </tr>
        </tfoot>
        {cartProducts}
      </table>
    );
  }
}
export default Cart;
