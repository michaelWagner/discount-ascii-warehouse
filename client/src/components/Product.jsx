import React from 'react';
import Cart from './Cart';

class Product extends React.Component {

  formatDateInRelativeTime(date) {
    let result = '';
    const timeAddedInMsec = Date.parse(date) / 1000;
    const currentTimeInMsec = (Date.parse(new Date())) / 1000;
    const timeRemainingInMsec = currentTimeInMsec - timeAddedInMsec;
    const unit = {
      'week':   Math.floor( timeRemainingInMsec / 604800 % 52 ),
      'day':    Math.floor( timeRemainingInMsec / 86400 % 7 ),
      'hour':   Math.floor( timeRemainingInMsec / 3600 % 24 ),
      'minute': Math.floor( timeRemainingInMsec / 60 % 60 ),
      'second': Math.floor( timeRemainingInMsec % 60 )
    };

    if (unit['week'] < 1) {
      for ( var key in unit ) {
        // Pluralize units for readability.
        if ( unit[key] > 1 ) {
          result += unit[key] + ' ' + key + 's ';
        }
        if ( unit[key] === 1 ) {
          result += unit[key] + ' ' + key + ' ';
        }
      }
      result += 'ago.'
    } else {
      result = date;
    }

    return result;
  }

  handleClick(event) {
    event.preventDefault();
    this.props.toggleCart(this.props.product);
  }

  styleFace(size) {
    return {'fontSize': size + 'px'};
  }

  render() {
    return (
      <tr key={this.props.product.id} onClick={this.handleClick.bind(this)}>
        <td>{this.props.product.id}</td>
        <td>{this.props.formatPrice(this.props.product.price)}</td>
        <td>{this.props.product.size}</td>
        <td style={this.styleFace(this.props.product.size)}>{this.props.product.face}</td>
        <td>{this.formatDateInRelativeTime(this.props.product.date)}</td>
        {this.props.parentComponent === "Cart" ? <td> x </td> : <td> + </td>}
      </tr>
    );
  }
}
export default Product;
