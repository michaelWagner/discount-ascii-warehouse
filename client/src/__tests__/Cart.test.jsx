import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Product from '../components/Product';
import Cart from '../components/Cart';

describe(Cart, () => {

  const products = [<Product />,<Product />];

  const totalPrice = products.reduce((total, product) => {
    return total + product.price;
  }, 0);
  const cartVisible = true;
  const mockFormatPriceFromCentsToDollars = jest.fn((price) => "$6.29");
  const mockToggleProductInCart = jest.fn();
  const component = shallow(
    <Cart products={products}
          totalPrice={totalPrice}
          toggleProductInCart={mockToggleProductInCart}
          formatPriceFromCentsToDollars={mockFormatPriceFromCentsToDollars}
          cartVisible={cartVisible}/>
  );

  it('renders and matches our snapshot', () => {
    const component = renderer.create(
      <Cart products={products}
            totalPrice={totalPrice}
            toggleProductInCart={mockToggleProductInCart}
            formatPriceFromCentsToDollars={mockFormatPriceFromCentsToDollars}
            cartVisible={cartVisible}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains the supplied Products', () => {
    expect(component.text()).toContain('<Product />');
  });

  it('displays the supplied total price', () => {
    expect(component.text()).toContain('Total price: $6.29');
  });

  it('formats the Carts price', () => {
    expect(mockFormatPriceFromCentsToDollars).toBeCalled();
  });
});
