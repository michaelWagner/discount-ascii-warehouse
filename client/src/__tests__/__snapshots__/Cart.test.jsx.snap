exports[`class Cart extends /* istanbul ignore next */_react2.default.Component {
  constructor() {
    super();
    this.state = {
      windowWidth: '0',
      colSpanWidth: 4 };

    this.updateWindowWidth = this.updateWindowWidth.bind(this);
  }

  componentDidMount() {
    this.updateWindowWidth();
    window.addEventListener('resize', this.updateWindowWidth);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.updateWindowWidth);
  }

  // Dynamically resize the cart depending on window width.
  updateWindowWidth() {
    let colSpanWidth = 4;
    const windowWidth = window.innerWidth;

    if (windowWidth <= 768) {
      colSpanWidth = 2;
    }

    this.setState({
      windowWidth: window.innerWidth,
      colSpanWidth: colSpanWidth });

  }

  render() {
    const cartProducts = this.props.products.map((product, idx) => {
      return (
        /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tbody', /* istanbul ignore next */{ key: idx, __source: { fileName: _jsxFileName, lineNumber: 41 }, __self: this },
          /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_Product2.default, /* istanbul ignore next */{ key: product.id, toggleProductInCart: this.props.toggleProductInCart,
            product: product, formatPriceFromCentsToDollars: this.props.formatPriceFromCentsToDollars,
            parentComponent: "Cart", __source: { fileName: _jsxFileName, lineNumber: 42 }, __self: this })));


    });

    return (
      /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'table', /* istanbul ignore next */{ className: (this.props.cartVisible ? "cart-table" : "cart-hidden") + " product-grid-table", __source: { fileName: _jsxFileName, lineNumber: 50 }, __self: this },
        /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tfoot', /* istanbul ignore next */{ className: 'cart-table-footer', __source: { fileName: _jsxFileName, lineNumber: 51 }, __self: this },
          /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tr', /* istanbul ignore next */{ className: 'table-footer', __source: { fileName: _jsxFileName, lineNumber: 52 }, __self: this },
            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'td', /* istanbul ignore next */{ colSpan: '2', __source: { fileName: _jsxFileName, lineNumber: 53 }, __self: this }, /* istanbul ignore next */'Total price: '),
            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'td', /* istanbul ignore next */{ className: 'table-footer-price', colSpan: this.state.colSpanWidth, __source: { fileName: _jsxFileName, lineNumber: 54 }, __self: this }, this.props.formatPriceFromCentsToDollars(this.props.totalPrice)))),


        cartProducts));


  }} renders and matches our snapshot 1`] = `
<table
  className="cart-table product-grid-table">
  <tfoot
    className="cart-table-footer">
    <tr
      className="table-footer">
      <td
        colSpan="2">
        Total price: 
      </td>
      <td
        className="table-footer-price"
        colSpan={4}>
        $6.29
      </td>
    </tr>
  </tfoot>
  <tbody>
    <tr
      onClick={[Function]}>
      <td
        style={
          Object {
            "fontSize": "undefinedpx",
          }
        } />
      <td />
      <td />
      <td>
        $6.29
      </td>
      <td />
      <td
        className="add-to-cart-btn">
        x
      </td>
    </tr>
  </tbody>
  <tbody>
    <tr
      onClick={[Function]}>
      <td
        style={
          Object {
            "fontSize": "undefinedpx",
          }
        } />
      <td />
      <td />
      <td>
        $6.29
      </td>
      <td />
      <td
        className="add-to-cart-btn">
        x
      </td>
    </tr>
  </tbody>
</table>
`;
